### ISSUE TITLE
unable to complete CallFunction() with low instruction count

### ISSUE THREAD

**Issue by incanus** [JamesBoer/Jinx#20]
Thanks for Jinx — I love it and have been evaluating it for some time now. Incidentally, I have a number of typo corrections (both in code and the docs) that I'll try to submit sometime in the near future. 

I'm working on a REPL of sorts using Jinx, so this might be a bit of an unusual case. I saw #19 which I thought would get me closer, but it's not quite the same situation. 

Because I have a REPL that I want to be interruptible (with `Esc`), I have set a low `globalParams.maxInstructions`, to the tune of 10 or so. That way, when the user wants to interrupt, not very many additional instructions are executed before the keyboard buffer is checked, the request is seen, and the script is cancelled. With a higher number, the interrupting is less responsive since the `script->Execute()` call will run for a longer time. 

Often, the scripts are procedural and straightforward. Other times, I have a use case where the user _can_ (not required) define the (public) function `tick()` which, if found in a script, will be executed repeatedly in between double-buffered screen swaps and vertical sync waits. This is after the whole script is "executed", i.e. after this ticking environment is fully setup. 

While I can check that the _whole script_ is fully parsed and executed (and thus `tick()` is registered if present) by calling `script->Execute()` repeatedly and checking for `script->IsFinished()`, with the low instruction count, the later calls to `CallFunction()` on the found `tick()` don't necessarily run the whole function. That is, if the body of `tick()` is more instructions than e.g. 10, it will never complete. 

Checking `script->IsFinished()` isn't appropriate here, since it is always "finished" at this point, and calling `script->Execute()` isn't either since it's already been executed fully. Also, the runtime parameters aren't meant to be changed after initial setup, so the instruction count can't be changed (and even if it could, making this a higher number would defeat the responsive cancelation needs mentioned above). 

Do you have any tips or thoughts about this? I can't see a way in the current code that I should be accomplishing this, but I may be missing something. I feel like something akin to `script->IsFinished()` but for function bodies would be useful.

**Comment by incanus** [comment]
Perhaps the REPL is confusing the topic here… I’m only ever executing a complete, parseable script. Sometimes it’s done in snippets entered interactively, (say, a `loop … end` section that runs when the scope is closed) and sometimes it’s a more involved script that has been edited in a visual editor, then run. So, for the purposes of this topic, it’s a complete script text that has `Execute()` called on it, but with a low instruction count, doesn’t always complete before the keyboard is checked for `Esc`. If it is _not_ pressed, another `Execute()` is run, which continues on with the script. 

In the case of a found `tick()` function, once the script is finished, the environment then moves from a repeated calling of `Execute()` to a repeated calling of `tick()`. In the case of no `tick()` found, the environment just exits once the script is finished. 

I’ll try to create a small program which better demonstrates this.

**Comment by JamesBoer** [comment]
Okay, so maybe it's better described as a mechanism to dynamically edit scripts on the fly?  If so, that makes a bit more sense to me.  Maybe the introduction of a debugging interface could help matters here.  

Anyhow, let me know if you can put together a test program, and I'll start investigating possible solutions.

**Comment by JamesBoer** [comment]
So, I'm kind of wondering if a coroutine would actually work for you.  Have you given it a try?  The fundamental problem is that CallFunction() is expecting the execution to success in a single Execute() cycle.  That's a problem i should address, but honestly, I may just have to signal a runtime error, because given the current API, there's no easy way to deal with that.

I just experimented in setting my unit tests to a very low instruction count (and I had to fix up a bunch of tests to account for this).   The one test that failed was the CallFunction() test.  But when I changed the code to create and call the function as a coroutine, it worked fine.  Internally, there's actually no difference between a function's execution being interrupted by a wait keyword or being interrupted by hitting the max instruction count without the error flag set.  So it just executes until it's finished.

As such, it seems like you should be able to just acquire a coroutine and use it in place of the IScript::CallFunction() API.  Could you let me know if that works?

**Comment by incanus** [comment]
Funny thing — I spent this afternoon on a plane, offline, and came to the same conclusion after some messing around. This is the key: 

>  Internally, there's actually no difference between a function's execution being interrupted by a wait keyword or being interrupted by hitting the max instruction count without the error flag set. 

I think the latter part being implicit in the tutorial is what kept me in the dark. That and I've been working on a late 2021 checkout of the code and only recently pulled in the async API changes, so they aren't as familiar to be as the rest of the API. 

This new bit in the tutorial set me on the experimentation path: 

> `ICoroutine::IsFinished()` checks to see if the coroutine is finished, and it not, executes until it is finished or another `wait` is encountered. 

First I tried with `wait` in my script (non-ideal, since I don't want the user to have to insert that) and then I realized it worked without it. 

I'll confirm for sure with some more testing, then close this out, but in the meantime, thanks for your followups. 

Incidentally, I didn't quite get there with a test code sample since I had a hard time replicating my actual setup in a way that produced a difference between something like `10` or `2000` max instructions. Long story short, I've got a lot of custom things going on, in a bare metal (writing my own kernel) environment, on specialized hardware, in case all of this sounds bonkers and somewhat unusual. It kind of is! I'll certainly share more when I can...
